### PR TITLE
Documentation for python requires

### PR DESCRIPTION
**Pull request by jgsogo**
Docs for: https://github.com/conan-io/conan/pull/4321

Examples: https://github.com/conan-io/examples/pull/9

### CODE REVIEW

File: extending/python_requires.rst [hunk 1]
```diff
@@ -1,160 +1,245 @@
...
+============================================
 
 .. warning::
 
     This is an **experimental** feature subject to breaking changes in future releases.
 
-The ``python_requires()`` feature allows to extend the recipes reusing Python code from other **conanfile.py** easily, even for inheritance
-approaches. The code to be reused will be in a *conanfile.py* recipe, and will be managed as any other Conan package.
+The ``python_requires()`` feature is a very convenient way to share files and code between
+different recipes. A *Python Requires* is just like any another recipe, it is the way it is
```
Comment (danimtb) [inline extending/python_requires.rst]:
```suggestion
different recipes. A *Python Requires* is just like any other recipe, it is the way it is
```

File: extending/python_requires.rst [hunk 2]
```diff
@@ -1,160 +1,245 @@
...
+            version = "version"
 
-It is not necessary to "create" any package binaries, or to ``upload --all``, because there are no binaries for this recipe.
+            exports = "scm_utils.py"
+            exports_sources = "CMakeLists.txt"
 
-Now, using the ``python_requires()`` we can write a new package recipe like:
 
-.. code-block:: python
+        def get_conanfile():
```
Comment (danimtb) [inline extending/python_requires.rst]: I would like to have this confirmed as the canonical/recommended way of using python requires

File: extending/python_requires.rst [hunk 3]
```diff
@@ -1,160 +1,245 @@
...
+
+            class BaseConanFile(ConanFile):
+
+                settings = "os", "compiler", "build_type", "arch"
+                options = {"shared": [True, False]}
+                default_options = "shared=False"
```
Comment (danimtb) [inline extending/python_requires.rst]:
```suggestion
                default_options = {"shared": False}
```

File: extending/python_requires.rst [hunk 4]
```diff
@@ -1,160 +1,245 @@
...
+      even consumers
+
+
+Reuse python sources
+--------------------
+
+In the example proposed we are using two functions through the ``base``
+variable: ``base.get_conanfile()`` and ``base.get_version()``. The first one is defined
+directly in the *conanfile.py* file, but the second one is in a different source file that
+we have exported together with the ``pyreq/version@user/channel`` recipe using the
```
Comment (danimtb) [inline extending/python_requires.rst]:
```suggestion
was exported together with the ``pyreq/version@user/channel`` recipe using the
```

File: extending/python_requires.rst [hunk 5]
```diff
@@ -1,160 +1,245 @@
...
-    :caption: **mydata.py**
 
-    src = "My cool source!"
-    build = "My cool build!"
-    pkg = "My cool package!"
-    info = "My cool package_info!"
+Reuse source files
+------------------
+
+Up to now, we have been reusing python code, but we can also package files within the
```
Comment (danimtb) [inline extending/python_requires.rst]: Really? You are kind of reusing the *scm_utils.py* in the previous example...

File: reference/conanfile/attributes.rst [hunk 6]
```diff
@@ -1172,3 +1171,41 @@ be computed. It can take three different values:
...
+
+    This attribute is part of the :ref:`python requires<python_requires>` feature, so
+    it is also an **experimental** feature subject to breaking changes in future releases.
+
+Python requires are associated with the ``ConanFile`` declared in the recipe file, data
+from those imported recipes is accessible using the ``python_requires`` attribute in
+the recipe itself. This attribute is a dictionary where the key is the name of the
+*python requires* reference and the value is a dictionary with the following information:
+
+ - ``ref``: full reference of the python requires
```
Comment (danimtb) [inline reference/conanfile/attributes.rst]:
```suggestion
 - ``ref``: Full reference of the python requires.
```

File: reference/conanfile/attributes.rst [hunk 7]
```diff
@@ -1172,3 +1171,41 @@ be computed. It can take three different values:
...
+ - ``exports_folder``: directory in the cache where the exported files are located.
```
Comment (danimtb) [inline reference/conanfile/attributes.rst]:
```suggestion
 - ``exports_folder``: Directory in the cache where the exported files are located.
```

File: extending/python_requires.rst (hunk 5 above)
Comment (jgsogo) [reply extending/python_requires.rst]: :unamused: ...for a C++ developer, python code are not **sources** 🤣 

I understand your concern, maybe it is the difference between `exports` and `exports_sources`, up to here I've been reusing code from the `exports` family, now I'm targeting the `exports_sources` files. If you can suggest names for those sections...

File: reference/conanfile/attributes.rst (hunk 6 above)
Comment (jgsogo) [reply reference/conanfile/attributes.rst]: After `:` it goes lower case, except if it is a quote...

File: reference/conanfile/attributes.rst (hunk 7 above)
Comment (jgsogo) [reply reference/conanfile/attributes.rst]: After `:` always lower case, except if it is a quote...

File: extending/python_requires.rst (hunk 2 above)
Comment (jgsogo) [reply extending/python_requires.rst]: Yes, this is something we have to agree, @conan-io/barbarians

File: extending/python_requires.rst (hunk 5 above)
Comment (danimtb) [reply extending/python_requires.rst]: ok, np

**Comment by danimtb** [comment]
We need to move this PR forward as it has relevant information for the 1.15 release

**Pull request merged**